taskcluster-pulse integration

Hexcles opened the pull request:
## Abstract

`PulseListener` is modified to support fetching Pulse credentials from [taskcluster-pulse](https://github.com/taskcluster/taskcluster-pulse/) with full backwards compatibility. Now, when creating a `PulseListener` instance, one can pass in the desired (Pulse) namespace and contact information (see the comments for detailed doc) instead of manually-configured Pulse username/password. `PulseListener` will then get Pulse credentials with necessary permissions from `taskcluster-pulse` before connecting, provided that the client has the correct scope. This process is transparent and the usage of constructed `PulseListener` instances is unaffected.

Also note that we will be using a new RabbitMQ name prefix (`taskcluster/{queues|exchanges}/...`, see L330 of `lib/pulselistener.js`). This change is worth some more discussion. I will drop by IRC on Monday to talk about it.

## Testing

Since only the code for establishing Pulse connection was changed, tests are focusing on this part too. Effectively, there are three supported methods to set up a connection now: 1. connection string; 2. Pulse username/password; 3. namespace and contact info (i.e. the new `taskcluster-pulse` way). We test the `bind` operation via these three methods individually.

When testing the `taskcluster-pulse` method, we mock its API endpoint using `nock` and return a fake response with Pulse username/password from `user-config.yml`, which was (and still is) precisely for testing purpose.

## Notes

* Currently `taskcluster-pulse` service is not working because of cert issues (taskcluster/taskcluster-pulse#60). Besides, it is managing a separate RabbitMQ cluster isolated from the production.
* I haven't done real integrating tests (partly of because of the issue above).
* Again, the top principle of this change is not to break any existing code.
* To migrate to `taskcluster-pulse`, all existing users of `PulseListener` must be modified to pass in the new constructor parameters. A TODO is to give deprecation warnings for hard-configured username/password.

Suggestions are more than welcome!

jonasfj commented on lib/pulselistener.js:
```diff
@@ -51,12 +52,29 @@ var retryConnect = function(connectionString, retries) {
+ *     payload: {
+ *       address:
+ *     }
+ *   }
```
You're missing a property: `credentials` with taskcluster credentials to passed to `new tc.Pulse({credentials})`.

I would also include an optional `baseUrl` for `new tc.Pulse({credentials, baseUrl})`, so it can be configured to work with taskcluster-pulse running locally.

jonasfj commented on lib/pulselistener.js:
```diff
@@ -51,12 +52,29 @@ var retryConnect = function(connectionString, retries) {
+ *                      // pulse.mozilla.org)
+ *   // Contact info sent to taskcluster-pulse when requesting credentials
+ *   // (required when credentials are managed by taskcluster-pulse)
+ *   contact: {         // One of:
```
Could we default this to:
```js
contact: {
  method: 'email',
  payload: {
    address: process.env.OWNER_EMAIL,
  },
}
```

We currently have a convention that we set the environment variable `OWNER_EMAIL` so we can easily see who owns a service...

If you rather call the env var something else we can also rename it...

@imbstack, thoughts?

jonasfj commented on lib/pulselistener.js:
```diff
@@ -65,12 +83,21 @@ var PulseConnection = function(options) {
-  if (!options.connectionString) {
-    options.connectionString = buildPulseConnectionString(options);
-  } else {
+  this.isTaskclusterPulse = false;
```
Prefix `_` to indicate that it's private.

Hexcles replied:
It's actually not private at the moment. See [L330](https://github.com/taskcluster/taskcluster-client/pull/64/files#diff-bdeec392e02b8fc5c8a9ac0b62f3b398R330), where `PulseListener` checks the `isTaskclusterPulse` property of `PulseConnection`. Is it better to also create this property in `PulseListener` and make both private?

jonasfj replied:
hmm... I somewhat regard that as private... since it shouldn't be read by someone using this code...

Perhaps it would be better to make a `queuePrefix` property instead? Similar to how we talked about it should work in `pulse-publisher`...

jonasfj commented on lib/pulselistener.js:
```diff
@@ -65,12 +83,21 @@ var PulseConnection = function(options) {
+    assert(options.contact,
+      "Must provide `contact` when Pulse credentials are managed by taskcluster-pulse");
+    this._contact = options.contact;
+    this._pulse = new tc.Pulse();
```
You don't need to keep reference to this object, I would create it when I need... especially as this code won't need it often...

jonasfj commented on lib/pulselistener.js:
```diff
@@ -51,12 +52,29 @@ var retryConnect = function(connectionString, retries) {
- *   connectionString:  // connectionString cannot be used with username,
- *                      // password and/or hostname.
+ *   hostname:          // Hostname to connect to (optional, defaults to
+ *                      // pulse.mozilla.org)
```
Ignored, if configured with `contact` and `credentials` are provided...

jonasfj commented on lib/pulselistener.js:
```diff
@@ -112,7 +140,21 @@ PulseConnection.prototype.connect = function() {
+          });
+        });
+    }
+    Promise.resolve(getCredentials).then(function() {
```
You should declare `getCredentials` sooner,  so it's not out of scope when you use it...

In some nice future we want to use `let` instead of `var`

Hexcles replied:
Ahh, unintended use of local variable hoisting in JS. Good lesson.

Hexcles commented:
Hi @jonasfj , I've addressed all your review comments here and most of our discussions on IRC. Only one thing is left at the moment but will go into a separate PR - retrying connection directly w/o asking for credentials from `tc-pulse` again if creds haven't expired.

The most important change that I wish you to take a closer look is how I group the properties in `options.credentials` and make them exclusive (both in code and in docs). Originally, `connectionString` and `username`/`password` are already exclusive practically. I just made the new `tc-pulse` way also exclusive to both of them, so now we have three well-separated groups of properties (three ways to set TC creds).

Thanks!

jonasfj commented on lib/pulselistener.js:
```diff
@@ -49,14 +59,42 @@ var retryConnect = function(connectionString, retries) {
+ *                      // to https://pulse.taskcluster.net/v1)
+ *   // Contact info sent to taskcluster-pulse when requesting credentials
+ *   // (optional, defaults to process.env.OWNER_EMAIL)
+ *   contact: {         // One of:
```
Hmm... Just a thought here, but instead of: `contact: {method: 'email', payload: {...}}`
What if we did:
```js
credentials:  {...}   // Taskcluster credentials
baseUrl:       '...', // BaseUrl for taskcluster-pulse, defaults to https://pulse.taskcluster.net/v1
email:          '...', // Email address to contact with alerts (required unless 'contact' is given).
contact: {      // Contact payload for taskcluster-pulse (not required if email is given)
  method: 'irc',
  payload: {channel: '...', user: '....'},
}
```
Basically, providing a shortcut for providing the contact email, as that is probably the most common method of contact.

Hexcles replied:
Good idea. It'd be more suitable to go into another PR later I think as it's kinda independent and compatible with the current implementation.

Hexcles commented:
taskcluster/taskcluster-pulse#64 contains the change of `namespace` API responses (with `connectionString` added), so now we have all the pieces coming together. I've updated some comments and the tests to reflect the latest change. The last thing, however, is that we might need an update of built-in API references (and a version bump) once the PRs in tc-pulse are merged. Cheers!

djmitche commented:
We decided to build a tc-lib-pulse instead, so this is not tied to tc-client.
